pkg/api/subscribeWorker.go

```diff
@@ -0,0 +1,196 @@
+	"go.uber.org/zap"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	subscriptionBufferSize    = 1024
+	maxSubscriptionsPerClient = 10000
+	SubscribeWorkerPollTime   = 100 * time.Millisecond
+	subscribeWorkerPollRows   = 10000
+	maxTopicLength            = 128
+)
+
+type listener = chan<- []*message_api.OriginatorEnvelope
+
+// A worker that listens for new envelopes in the DB and sends them to subscribers
+// Assumes that there are many listeners - non-blocking updates are sent on buffered channels
+// and may be dropped if full
+type subscribeWorker struct {
+	ctx context.Context
+	log *zap.Logger
+
+	dbSubscription <-chan []queries.GatewayEnvelope
+	// Assumption: listeners cannot be in multiple slices
+	globalListeners     []listener
+	originatorListeners map[uint32][]listener
+	topicListeners      map[string][]listener
+}
+
+func startSubscribeWorker(
+	ctx context.Context,
+	log *zap.Logger,
+	store *sql.DB,
+) (*subscribeWorker, error) {
+	log = log.Named("subscribeWorker")
+	q := queries.New(store)
+	pollableQuery := func(ctx context.Context, lastSeen db.VectorClock, numRows int32) ([]queries.GatewayEnvelope, db.VectorClock, error) {
+		envs, err := q.
+			SelectGatewayEnvelopes(
+				ctx,
+				*db.SetVectorClock(&queries.SelectGatewayEnvelopesParams{}, lastSeen),
+			)
+		if err != nil {
+			return nil, lastSeen, err
+		}
+		for _, env := range envs {
+			// TODO(rich) Handle out-of-order envelopes
+			lastSeen[uint32(env.OriginatorNodeID)] = uint64(env.OriginatorSequenceID)
+		}
+		return envs, lastSeen, nil
+	}
+
+	vc, err := q.SelectVectorClock(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	subscription := db.NewDBSubscription(
+		ctx,
+		log,
+		pollableQuery,
+		db.ToVectorClock(vc),
+		db.PollingOptions{
+			Interval: SubscribeWorkerPollTime,
+			NumRows:  subscribeWorkerPollRows,
+		},
+	)
+	dbChan, err := subscription.Start()
+	if err != nil {
+		return nil, err
+	}
+	worker := &subscribeWorker{
+		ctx:                 ctx,
+		log:                 log,
+		dbSubscription:      dbChan,
+		globalListeners:     make([]listener, 0),
+		originatorListeners: make(map[uint32][]listener),
+		topicListeners:      make(map[string][]listener),
+	}
+
+	go worker.start()
+
+	return worker, nil
+}
+
+func (s *subscribeWorker) start() {
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case new_batch := <-s.dbSubscription:
+			for _, row := range new_batch {
+				s.dispatch(&row)
+			}
+		}
+	}
+}
+
+func (s *subscribeWorker) dispatch(
+	row *queries.GatewayEnvelope,
+) {
+	bytes := row.OriginatorEnvelope
+	env := &message_api.OriginatorEnvelope{}
+	err := proto.Unmarshal(bytes, env)
+	if err != nil {
+		s.log.Error("Failed to unmarshal envelope", zap.Error(err))
+		return
+	}
+	for _, listener := range s.originatorListeners[uint32(row.OriginatorNodeID)] {
+		select {
+		case listener <- []*message_api.OriginatorEnvelope{env}:
```

Comment:
making a comment on this line, but it applies to all.

This feels like a very easy denial of service attack. If any channel/subscriber can't keep up, none will receive messages.

Am I reading the code correctly?

Reply:
The `select{}` with `default` is a non-blocking write to the buffered channel, so we don't depend on subscribers to keep up

Reply:
Gotcha. So as the code is written in the PR; if the channel is full we don't broadcast more. This will lead to gaps I think. The eventual plan is to kill listeners that are not keeping up?